### ISSUE TITLE
Kerberos 5, etype 18, Pre-Auth (19900) with salt

### ISSUE THREAD

**Issue by JasonMalik** [hashcat/hashcat#2783]
Hello,

Is it possible to add support for salt when using Kerberos 5, etype 18, Pre-Auth?

$krb5pa$18$user$domain$salt$token

I can crack such a format using John, but I'd rather use Hashcat.

Thank you.

**Comment by jsteube** [comment]
Maybe @brandoncasaba can jump in here?

**Comment by brandoncasaba** [comment]
Sure, the algorithm is supported as of Hashcat 6.0.0, but without a salt field. It looks like the newer John implementation lets you override the salt, but the default behavior is for the salt to be "realm.user" (or domain.user) which are already supplied values. This is actually the only thing the supplied realm/username fields are used for. This concatenated salt behavior is assumed and not configurable in either the older John implementation or the current Hashcat implementation. If there is an environment where this behavior does not work, I'd be curious to know. 

https://github.com/openwall/john/blob/8bcaddef35f703c47367fad612e473061459b156/src/opencl_krb5pa-sha1_fmt_plug.c

https://www.openwall.com/lists/john-users/2012/11/18/5/1

**Comment by jsteube** [comment]
I see, so it's a pure change only in the module, not the kernel?

**Comment by brandoncasaba** [comment]
Correct, this is where the salt, in this case it looks like the variable name is "account_info" is put together.
https://github.com/hashcat/hashcat/blob/master/src/modules/module_19900.c#L149-L176

**Comment by jsteube** [comment]
I think that's not so complicated task then. 

1. The difficulty would be to have a branch to decide which decoding path to go in the module, because in the end we want to support both input formats. Having two decoding paths in combination with input_tokenizer is no magic and it wouldn't be the first time we do this. A nice example on how to do this is `src/modules/module_09800.c`. 
2.  We will also need to change the encoding path depending on how the input was because we need to guarantee the output format matches the original input format. So we need to store the input format type in an additional field in the esalt. This will typically also require to update the kernel because the esalt structure is stored in two places, the module and the kernel. Even if we do not use it from the kernel it's important to update it, otherwise the length of the strucure is not sychronized.
3. Optionally the perl unit test also need an update to support both formats, but that's really optional.

@brandoncasaba or @JasonMalik anyone from you to make the change?

**Comment by JasonMalik** [comment]
I have never worked on Hashcat, I am an end user, but if there is anything I can to to help I will gladly do it.

**Comment by brandoncasaba** [comment]
I can take a look this evening, I don't imagine the LOE would be more than a couple hours with testing. 

Do you have samples w/ an included salt? It looks like the salt is a standard character array, not a hex encoded byte array. Is that correct?

I went back and looked to mode 7500 which is the old RC4 Kerberos Pre-auth and it looks like it requires a salt value, but it also takes a user/realm and does nothing with them. So it's probably better to support both formats and prefer using the salt when provided. As best as I can tell for mode 7500, it's simply expecting a string (non-hex) value for salt. 

The john implementation appears to have a max length of 52, I assume that's sufficient?

**Comment by jsteube** [comment]
I don't have any real world example. @JasonMalik can you provide the samples?

**Comment by bishwabikash** [comment]
Working on this. PR #4884 covers 19800 and 19900, built the way you described in this thread.
